### PR TITLE
I added functionality to symlink a "latest" directory to the latest snapshot.

### PR DESCRIPTION
**Pull request by Gooberpatrol66**

### CODE REVIEW

**Comment by plattfot** [comment]
Thanks for the PR.

There are a few things I would like you to change before I merge this feature.
First off, don't create a PR from your master branch, instead move this to it's own branch. It will save you a lot of headaches in the future.

1. `link-latest` pulls in one new dependencies `ls`. I would like to keep the dependencies to a minimum so please switch to `find`
```
function link-latest {
    local snapshots
    mapfile -t snapshots < <(find -maxdepth 1 -mindepth 1 -type  -printf "%f\n" | sort -k1,1nr)
    ln -s "$1/${snapshots[0]}" "$1/latest-tmp"
    mv -T "$1/latest-tmp" "$1/latest"
}
```
2. list-snapshots needs to be updated to not return `latest` to `baksnapper`. We should be able to refactor the find code to be used in both this and `link-latest`

3. There is no error handling if all snapshots are deleted and `link-latest` is called.

4. This functionality should be opt-in and exposed to `baksnapper` as a new API call in `baksnapperd`.

File: baksnapperd.sh [hunk 1]
```diff
@@ -38,6 +38,21 @@ function warning {
     echo -e "[Warning] $1" 1>&2
 }
 
+function link-latest {
+    arr=()
+    for dir in "$1"/*; do
+        if ! [[ $(echo "$dir" | grep "latest") ]]; then
```
Comment (plattfot) [inline baksnapperd.sh:44]: No need to shell out to just check if latest is in the backup directory. Also there is a subtile bug if the backup location for some reason contains latest e.g. `$1 == /my/latest/backup/`.
Instead you can do this with [parameter expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) see the section about ${parameter##word}
I.e.
```
if  [[ "${dir##*/}" != "latest" ]]
```

File: baksnapperd.sh [hunk 2]
```diff
@@ -38,6 +38,21 @@ function warning {
...
+            if [[ -d "$dir/snapshot" ]]; then
```
Comment (plattfot) [inline baksnapperd.sh:45]: You should be able to combine the above check with this
I.e.

```
if [[ "${dir}##*/" != "latest" && -d "$dir/snapshot"
```

File: baksnapperd.sh [hunk 3]
```diff
@@ -38,6 +38,21 @@ function warning {
     echo -e "[Warning] $1" 1>&2
 }
 
+function link-latest {
+    arr=()
```
Comment (plattfot) [inline baksnapperd.sh:42]: Use `declare -a` as that will make the array local to the function.
Also try and use more descriptive names than `arr`, for example `snapshots`.

File: latest-bash [hunk 4]
```diff
@@ -0,0 +1 @@
+/media/backup/nathan-gentoodesktop/6295
```
Comment (plattfot) [inline latest-bash:1]: Please remove private files from the merge request.
Thanks

**Comment by plattfot** [comment]
Looks good, I'll just do some testing end of this week to see if everything works then I'll merge this.

File: baksnapper.sh [hunk 5]
```diff
@@ -55,6 +55,8 @@ Options:
 --type TYPE       Specify either to backup snapshots to a server
                   (push) or to backup snapshots from a server
                   (pull). Default is to push.
+
```
Comment (plattfot) [inline baksnapper.sh:58]: Remove the whitespaces for this line

**Comment by plattfot** [comment]
Founds some bugs lurking in the code, here is the diff split into the bugs I found, the whole is at the end of this comment.

```diff
 --type TYPE       Specify either to backup snapshots to a server
                   (push) or to backup snapshots from a server
                   (pull). Default is to push.
-                  
+
 --link            Create a "latest" directory linked to the latest snapshot
 
 -v, --verbose     Verbose print out.
```
Remove trailing whitespaces
```diff
 
 # Use getopt to parse the command-line arguments
 
-if ! _args=$(getopt --name "baksnapper" --options "adhvp" --long "config:,configfile:,delete:,daemon:,private-key:,snapshot:,type:,all,delete-all,help,prune,verbose,version" -- "$@")
+if ! _args=$(getopt --name "baksnapper" --options "adhvp" --long "config:,configfile:,delete:,daemon:,private-key:,snapshot:,type:,all,delete-all,help,prune,verbose,version,link" -- "$@")
 then
     error "Try '$0 --help for more information.'"
 fi
```
Missing `link` in the long option argument. Did you forget to commit this?
```diff
         $receiver remove-broken-snapshot "$dest_root" "$1"
         error "Failed to send snapshot!"
     fi
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
@@ -589,7 +596,7 @@ then
         case $answer in
             y|Y)
                 $receiver remove-snapshots "$dest_root" "${dest_snapshots[@]}"
-                if [ $p_link -eq 1 ]
+                if [[ $p_link -eq 1 ]]
                 then
                     $receiver link-latest "$dest_root"
                 fi
@@ -607,7 +614,7 @@ then
     backup
 else
     $receiver remove-snapshots "$dest_root" "${p_delete_list[@]}"
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
@@ -616,7 +623,7 @@ fi
 if [[ ${p_prune-0} == 1 ]]
 then
     $receiver remove-snapshots "$dest_root" "${only_in_dest[@]}"
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
``` 
The `$p_link` will expand to nothing when not set, which it will do if `--link` is not set, causing a syntax error. Either quote them (`"$p_link"`) or use `[[` as I did in the diff.

```diff
@@ -133,7 +133,8 @@ case "$1" in
         if ! [ ${#snapshots[@]} -eq 0 ]; then
             ln -sfn "${snapshots[-1]}" "$1/latest-tmp"
             mv -T "$1/latest-tmp" "$1/latest"
-        else
+        elif [[ -e "$1/latest" ]]
+        then
             rm "$1/latest"
         fi
         ;;
```
Will error out when running `--delete-all` and no `latest` exist.


```diff
diff --git a/baksnapperd.sh b/baksnapperd.sh
index 67147ca..91b3ff3 100755
--- a/baksnapperd.sh
+++ b/baksnapperd.sh
@@ -40,7 +40,7 @@ function warning {
 
 case "$1" in
     version) # Return what version of the API it's using, always one integer
-        echo 2
+        echo 3
         ;;
     list-snapshots) # List snapshots at backup location
         shift
```
Bump up the version in baksnapperd to `3` as the API has changed. 
```diff
@@ -355,19 +355,26 @@ then
     sender_version=1
 fi
 
-if [[ "$receiver_version" -gt 2 ]]
+if [[ "$receiver_version" -gt 3 ]]
 then
     error "receiver is too new, need to use version 2 or 1"
 fi
 
-if [[ "$sender_version" -gt 2 ]]
+if [[ "$sender_version" -gt 3 ]]
 then
     error "sender is too new, need to use version 2 or 1"
 fi
 
+
+if [[ $receiver_version -lt 3 && $p_link -eq 1 ]]
+then
+    warning "Receiver does not support '--link', turning it off"
+    p_link=0
+fi
+
 # Get the subvolume to backup
 case $sender_version in
-    2)
+    2|3)
         subvolume=$($sender get-snapper-root "$p_config")
         ;;
     1)
@@ -388,7 +395,7 @@ $receiver create-config "$dest_root" || error "Problem creating config at backup
 
 # List all the snapshots available
 case $sender_version in
-    2)
+    2|3)
         mapfile -t src_snapshots < <($sender list-snapshots "$src_root")
         ;;
     1)
@@ -402,7 +409,7 @@ num_src_snapshots=${#src_snapshots[@]}
 
 # List all the snapshots at the backup location
 case $receiver_version in
-    2)
+    2|3)
         mapfile -t dest_snapshots < <($receiver list-snapshots "$dest_root")
         ;;
     1)
```
Update `baksnapper` to handle the new version of the API.

Also don`t forget to add 
```
# Copyright (C) 2021  YOUR NAME <EMAIL>
```
Under my name for `baksnapper.sh` and `baksnapperd.sh`. I noticed I missed to add my email, but will do that after this has merged to avoid unnecessary merge conflicts. Replace `YOUR NAME` with the name you get from 
```bash
git config --global user.name
```
and `EMAIL` with 
```bash
git config --global user.email
```

Here is the full diff:
```diff
diff --git a/baksnapper.sh b/baksnapper.sh
index fc1e031..fa55d75 100755
--- a/baksnapper.sh
+++ b/baksnapper.sh
@@ -55,7 +55,7 @@ Options:
 --type TYPE       Specify either to backup snapshots to a server
                   (push) or to backup snapshots from a server
                   (pull). Default is to push.
-                  
+
 --link            Create a "latest" directory linked to the latest snapshot
 
 -v, --verbose     Verbose print out.
@@ -188,7 +188,7 @@ function read-config {
 
 # Use getopt to parse the command-line arguments
 
-if ! _args=$(getopt --name "baksnapper" --options "adhvp" --long "config:,configfile:,delete:,daemon:,private-key:,snapshot:,type:,all,delete-all,help,prune,verbose,version" -- "$@")
+if ! _args=$(getopt --name "baksnapper" --options "adhvp" --long "config:,configfile:,delete:,daemon:,private-key:,snapshot:,type:,all,delete-all,help,prune,verbose,version,link" -- "$@")
 then
     error "Try '$0 --help for more information.'"
 fi
@@ -480,7 +480,7 @@ function single-backup {
         $receiver remove-broken-snapshot "$dest_root" "$1"
         error "Failed to send snapshot!"
     fi
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
@@ -589,7 +589,7 @@ then
         case $answer in
             y|Y)
                 $receiver remove-snapshots "$dest_root" "${dest_snapshots[@]}"
-                if [ $p_link -eq 1 ]
+                if [[ $p_link -eq 1 ]]
                 then
                     $receiver link-latest "$dest_root"
                 fi
@@ -607,7 +607,7 @@ then
     backup
 else
     $receiver remove-snapshots "$dest_root" "${p_delete_list[@]}"
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
@@ -616,7 +616,7 @@ fi
 if [[ ${p_prune-0} == 1 ]]
 then
     $receiver remove-snapshots "$dest_root" "${only_in_dest[@]}"
-    if [ $p_link -eq 1 ]
+    if [[ $p_link -eq 1 ]]
     then
         $receiver link-latest "$dest_root"
     fi
diff --git a/baksnapperd.sh b/baksnapperd.sh
index 67147ca..bdbc7af 100755
--- a/baksnapperd.sh
+++ b/baksnapperd.sh
@@ -133,7 +133,8 @@ case "$1" in
         if ! [ ${#snapshots[@]} -eq 0 ]; then
             ln -sfn "${snapshots[-1]}" "$1/latest-tmp"
             mv -T "$1/latest-tmp" "$1/latest"
-        else
+        elif [[ -e "$1/latest" ]]
+        then
             rm "$1/latest"
         fi
         ;;
```

**Comment by plattfot** [comment]
I've added my email to the headers so you have a better example on how to add the credit. Please, rebase your branch on my master branch.
And please use a valid email address or if you don't feel comfortable giving one just give your name, that is fine.
I'll do some testing later this week and see if everything works and hopefully no more bugs shows up so this can be merged!

**Comment by plattfot** [comment]
Found a bug which I introduced in the previous patch, for some reason `-e` is not valid for a symlink so it will not clean up the `latest` directory when all snapshots are removed. Switching to checking if it exist and is a symlink `-h` fixes the issue.

Also there's a trailing whitespace after your name in `baksnapperd.sh`.
```diff
diff --git a/baksnapperd.sh b/baksnapperd.sh
index 1fe47a0..4becdba 100755
--- a/baksnapperd.sh
+++ b/baksnapperd.sh
@@ -3,7 +3,7 @@
 # Baksnapperd - Daemon used by baksnapper when backup via ssh
 
 # Copyright (C) 2015-2021  Fredrik Salomonsson <plattfot@posteo.net>
-# Copyright (C) 2021  Nathan Dehnel 
+# Copyright (C) 2021  Nathan Dehnel
 
 # This file is part of baksnapper
 
@@ -132,7 +132,7 @@ case "$1" in
         if ! [ ${#snapshots[@]} -eq 0 ]; then
             ln -sfn "${snapshots[-1]}" "$1/latest-tmp"
             mv -T "$1/latest-tmp" "$1/latest"
-        elif [[ -e "$1/latest" ]]
+        elif [[ -h "$1/latest" ]]
         then
             rm "$1/latest"
         fi
```

**Comment by plattfot** [comment]
Forgot to mention, can you [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) your changes on top of mine instead of merging like you currently are doing? Makes the history cleaner.
Let me know if you need a small guide in how to do that as undo a merge can be a bit tricky.

Thanks.

**Comment by Gooberpatrol66** [comment]
>Let me know if you need a small guide in how to do that as undo a merge can be a bit tricky.
Yeah, that would be helpful.

**Comment by plattfot** [comment]
First do a backup of your repo, in case something goes wrong
```
cp -r baksnapper{,.bak}
```
Then cd into `baksnapper` (or the name you call your local clone).
Change to the branch `wip-link` and run
```
git log --oneline --graph -22
```
Look for the commit `add credit` as that is your last commit, my output looks something like this
```*   a5c2468 (HEAD -> wip-link, origin/wip-link) Merge branch 'master' into wip-link
|\  
| * 6fc965e Update copyright header for files
| * 9574b18 Add copy of gpl3 license
* | 4815f77 (origin/master, origin/HEAD, master) add credit
* | cb26dd4 apply patch
* | 4e4d5e5 remove variable
* | ef70ceb link-latest: add shift
* | e6924f8 backup-link: parse config
* | c6fbf58 remove extra function
* | 965c591 link-latest: remove link if no snapshots exist
* | b9b98d2 add link-latest to API
* | 7659cd8 list-snapshots: directories only
* | 3c9d3c4 simplify checks
* | fd13d3a change array name
* | 7a7bf7d prevent latest from linking to itself
* | 988784c overwrite symlink
* | 2b03163 only link if array is not empty
* | 0d17ad5 remove file
* | 8cea0e8 ensure latest does not link to a directory unless it contains a snapshot
* | fc63c56 add link-latest to remove-snapshots
* | 7daf2bc add function to create a latest symlink
|/  
* 2715350 (tag: 2.0.0) Bump up major: 1.0.2 -> 2.0.0
```
Reset the `wip-branch` to your last commit (the `4815f77 (origin/master, origin/HEAD, master) add credit`) using the checksum listed
```
git reset --hard 4815f77
```
Fetch the master from my repo, assuming you're using `upstream` as its remote name
```
git fetch upstream master
```
Then rebase the `wip-link` branch onto that using
```
git rebase upstream/master
```
It will fail with an error message:
```
Auto-merging baksnapperd.sh
CONFLICT (content): Merge conflict in baksnapperd.sh
Auto-merging baksnapper.sh
CONFLICT (content): Merge conflict in baksnapper.sh
error: could not apply 4815f77... add credit
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 4815f77... add credit
```
Resolve the conflicts for `baksnapperd.sh` and `baksnapper.sh` (similar to what you did when merging I would guess). Then run
```
git rebase --continue
```
Don't change the commit message when it pops up just accept it.

After you should have your changes on top of mine, to verify run `git log --oneline --graph -22` again
```
* 35e1c4c (HEAD -> wip-link) add credit
* 0c7852f apply patch
* 224b76a remove variable
* fe2126f link-latest: add shift
* 48068fa backup-link: parse config
* f4e5b3c remove extra function
* 7d767d2 link-latest: remove link if no snapshots exist
* f314753 add link-latest to API
* 44c14d4 list-snapshots: directories only
* 42551e9 simplify checks
* 4fb0a21 change array name
* 39387ab prevent latest from linking to itself
* 2757ec3 overwrite symlink
* 9176c34 only link if array is not empty
* bd26432 remove file
* 76b371e ensure latest does not link to a directory unless it contains a snapshot
* 4bdcc4f add link-latest to remove-snapshots
* 36e1795 add function to create a latest symlink
* 6fc965e (upstream/master) Update copyright header for files
* 9574b18 Add copy of gpl3 license
* 2715350 (tag: 2.0.0) Bump up major: 1.0.2 -> 2.0.0
* 0ff8a0c Update baksnapper to handle both API 1 and 2
```
Note the `Update copyright header for files` are now near the bottom of the commits.

All that is left is to rewind your master branch to track upstream again.
```
git checkout master
git reset --hard 2.0.0
git pull --rebase upstream/master
```
If everything looks good then you can push both your `wip-link` branch and `master` branch to `origin`. 
Note that you need to force push both branches
```
git push --force
```

**Comment by Gooberpatrol66** [comment]
```
nathan@gentoodesktop ~/Documents/projects/baksnapper $ git pull --rebase upstream/master
fatal: 'upstream/master' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
nathan@gentoodesktop ~/Documents/projects/baksnapper $ git remote add upstream https://github.com/plattfot/baksnapper.git
fatal: remote upstream already exists.
nathan@gentoodesktop ~/Documents/projects/baksnapper $ git remote -v
origin  https://github.com/Gooberpatrol66/baksnapper.git (fetch)
origin  https://github.com/Gooberpatrol66/baksnapper.git (push)
upstream        https://github.com/plattfot/baksnapper.git (fetch)
upstream        https://github.com/plattfot/baksnapper.git (push)

```
I got an error near the end of your guide.

**Comment by plattfot** [comment]
Sorry, typo on the `git pull` command. I've updated the original comment.
And for clarity it should be:
```
git pull --rebase upstream master
```
I.e. drop the `/` between the `upstream` and `master`.

**Comment by Gooberpatrol66** [comment]
Ok thanks, now test it out.

**Comment by plattfot** [comment]
I think you picked the wrong commit when you resolved the conflict. As I don't see your name in the `add credit` commit anymore, instead it restore some part of what I removed in `Update copyright header for files`. Please fix and do a force push to correct it.
You can either redo the rebase from your backup, or fix it in place and squash that commit onto the `add credit`.
Here's a blog post on how to squash: https://gitbetter.substack.com/p/how-to-squash-git-commits

Also don't forget to fix:
```diff
diff --git a/baksnapperd.sh b/baksnapperd.sh
index 0f07b4b..773b33b 100755
--- a/baksnapperd.sh
+++ b/baksnapperd.sh
@@ -133,7 +133,7 @@ case "$1" in
         if ! [ ${#snapshots[@]} -eq 0 ]; then
             ln -sfn "${snapshots[-1]}" "$1/latest-tmp"
             mv -T "$1/latest-tmp" "$1/latest"
-        elif [[ -e "$1/latest" ]]
+        elif [[ -h "$1/latest" ]]
         then
             rm "$1/latest"
         fi
```

**Comment by Gooberpatrol66** [comment]
Is it fixed now?

File: baksnapper.sh [hunk 6]
```diff
@@ -97,6 +100,8 @@ Copyright (C) 2015-2021  Fredrik Salomonsson
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
 This is free software: you are free to change and redistribute it.
 There is NO WARRANTY, to the extent permitted by law.
+
+Written by Fredrik "PlaTFooT" Salomonsson
```
Comment (plattfot) [inline baksnapper.sh:104]: Kind of, this is still here

File: baksnapperd.sh [hunk 7]
```diff
@@ -131,7 +134,7 @@ case "$1" in
         if ! [ ${#snapshots[@]} -eq 0 ]; then
             ln -sfn "${snapshots[-1]}" "$1/latest-tmp"
             mv -T "$1/latest-tmp" "$1/latest"
-        elif [[ -e "$1/latest" ]]
+        elif [[ -h "$1/latest" ]]
```
Comment (plattfot) [inline baksnapperd.sh:137]: Can you break this out into a separate commit? Instead of having it hide in the `add credit` commit.
Simplest is to just remove it and squash that into `add credit` as you've done with fixing the credit.
Then add it back and commit that as a separate commit.

Thanks

**Comment by Gooberpatrol66** [comment]
How about now?

File: baksnapperd.sh [hunk 8]
```diff
@@ -25,6 +26,8 @@ Copyright (C) 2015-2021  Fredrik Salomonsson
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
 This is free software: you are free to change and redistribute it.
 There is NO WARRANTY, to the extent permitted by law.
+
+Written by Fredrik "PlaTFooT" Salomonsson
```
Comment (plattfot) [inline baksnapperd.sh:30]: Missed one

**Comment by plattfot** [comment]
Also you might want to edit the commit message
``` 
add credit

fix rebase

fix rebase

revert
```
The "fix rebase" and "revert" are just noise.
Thanks

**Comment by Gooberpatrol66** [comment]
Ok

File: baksnapper.sh [hunk 9]
```diff
@@ -489,6 +505,10 @@ function incremental-backup {
         $receiver remove-broken-snapshot "$dest_root" "$1"
         error "Failed to send snapshot!"
     fi
+    if [ $p_link -eq 1 ]
```
Comment (plattfot) [inline baksnapper.sh:508]: Change this to `[[ $p_link -eq 1 ]]`
Getting `[: -eq: unary operator expected` when running without `--link` option

**Comment by plattfot** [comment]
Looks good, merged it.
Thank you for the feature!
